src/Functions/toStartOfInterval.cpp

```diff
@@ -243,15 +357,64 @@ class FunctionToStartOfInterval : public IFunction
         size_t size = time_data.size();
 
         auto result_col = result_type->createColumn();
-        auto *col_to = assert_cast<ToColumnType *>(result_col.get());
+        auto * col_to = assert_cast<ToColumnType *>(result_col.get());
         auto & result_data = col_to->getData();
         result_data.resize(size);
 
         Int64 scale_multiplier = DecimalUtils::scaleMultiplier<DateTime64>(scale);
+        Int64 scale_on_interval = decideScaleOnPrecision<unit>(scale);
```

Comment:
I don't really get why we need `decideScaleOnPrecision`, `scale_on_interval`, and `scale_diff`. Either please add some comments to the code or simplify.